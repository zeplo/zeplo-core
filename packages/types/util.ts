
export type Maybe<T> = T | null

export type JsonValue =  boolean | number | string | null | JsonArray | JsonMap
export interface JsonMap { [key: string]: JsonValue }
export interface JsonArray extends Array<JsonValue> {}

export type DeepPartial<T> = T extends Function ? T : (T extends object ? { [P in keyof T]?: DeepPartial<T[P]>; } : T);

/**
 * Get all of the keys to which U can be assigned.
 */
export type OnlyKeys<T, U> = {
  [K in keyof T]: U extends T[K] ? K : never
}[keyof T];

/**
 * Get the interface containing only properties to which U can be assigned.
 */
export type OnlyUndefined<T> = {
  [K in OnlyKeys<T, undefined>]: T[K]
}

/**
 * Get all of the keys except those to which U can be assigned.
 */
export type ExcludeKeys<T, U> = {
  [K in keyof T]: U extends T[K] ? never : K
}[keyof T];

/**
 * Get the interface containing no properties to which U can be assigned.
 */
export type ExcludeUndefined<T> = {
  [K in ExcludeKeys<T, undefined>]: T[K]
}

/**
 * Get the interface where all properties are optional.
 */
export type Optional<T> = {[K in keyof T]?: T[K]};

/**
 * Get the interface where properties that can be assigned undefined are
 * also optional.
 */
export type UndefinedOptional<T> = ExcludeUndefined<T> & Optional<OnlyUndefined<T>>

export type DeepUndefinedOptional<T> = T extends (...args: any[]) => any
  ? T
  : T extends any[]
  ? _DeepRequiredArray<T[number]>
  : T extends object
  ? _DeepRequiredObject<UndefinedOptional<T>>
  : T;

 export interface _DeepRequiredArray<T>
  extends Array<DeepUndefinedOptional<T>> {}

  export type _DeepRequiredObject<T> = {
  [P in keyof T]: DeepUndefinedOptional<T[P]>
}