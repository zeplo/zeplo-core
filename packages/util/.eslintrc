{
  "extends": ["plugin:@typescript-eslint/recommended"],
  "parser":  "@typescript-eslint/parser",
  "env": {
    "browser": true,
    "node": true,
    "es6": true,
    "mocha": true,
    "jest": true
  },
  "rules": {
    "semi": ["error", "never"],
    "no-underscore-dangle": 0,
    "space-before-function-paren": ["error", "always"],
    "arrow-body-style": 0,
    "no-use-before-define": ["error", { "functions": false }],
    "arrow-parens": 0,
  }
}
