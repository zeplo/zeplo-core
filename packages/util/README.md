# @zeplo/util

Utility functions that are used across packages.