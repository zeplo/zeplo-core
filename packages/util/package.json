{
  "name": "@zeplo/util",
  "version": "2.1.0",
  "main": "dist/index.js",
  "files": [
    "dist/*"
  ],
  "license": "MIT",
  "scripts": {
    "clean": "rimraf dist",
    "dev": "env-cmd ts-node",
    "build": "tsc",
    "prepublish": "yarn clean && yarn build"
  },
  "devDependencies": {
    "@types/express": "^4.17.13",
    "@types/fs-extra": "^9.0.13",
    "@types/lodash": "^4.14.178",
    "@types/node": "^17.0.18",
    "@types/uuid": "^8.3.4",
    "@typescript-eslint/eslint-plugin": "^5.12.0",
    "@typescript-eslint/parser": "^5.12.0",
    "env-cmd": "^10.1.0",
    "eslint": "^8.9.0",
    "rimraf": "^3.0.2",
    "ts-node": "^10.5.0",
    "ts-node-dev": "^1.1.8",
    "typescript": "^4.5.5"
  },
  "dependencies": {
    "@zeplo/errors": "^2.0.8",
    "@zeplo/types": "^2.1.0",
    "express": "^4.17.3",
    "express-async-handler": "^1.2.0",
    "lodash": "^4.17.21",
    "uuid": "^8.3.2"
  }
}
