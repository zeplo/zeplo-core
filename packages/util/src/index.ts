export { cleanQuery, cleanHeaders } from './clean'
export { 
  parseRequest, 
  parseBody, 
  getBackoff, 
  getStartTimeFromFeatures,
  getRawFeatures,
  parseRawRequestObject, 
  createRequestJob ,
  getValueFromQueryOrHeader,
} from './parse'
export { 
  createStepsJobs,
  hasCyclicalDependency,
} from './steps'
export { 
  requestHandler,
  jsonRequestHandler,
  queryToString,
  queryToNumber,
  parseQuery,
} from './request'