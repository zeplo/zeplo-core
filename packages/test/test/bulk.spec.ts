/**
 * @jest-environment node
 */
import qs from 'query-string'
import { omit } from 'lodash'
import { createWorkspace } from '#/util/data'
import waitFor from '#/util/waitFor'
import queue from '#/util/queue'

jest.setTimeout(30000)

describe('bulk.spec', () => {
  const workspace = `test-workspace-bulk-${Date.now()}`
  const token = `test-token-bulk-${Date.now()}`

  let cleanup: () => Promise<void>
  beforeAll(async () => {
    cleanup = await createWorkspace(workspace, token)
  })
  afterAll(() => cleanup())

  test('single bulk request with URL only', async () => {
    const res = await queue.post(`/bulk?_token=${token}`, [{
      url: 'https://postman-echo.com/post',
    }])
    expect(res.data).toHaveLength(1)
    expect(res.data[0]).toHaveProperty('id')
    const id = res.data[0].id

    const req = await waitFor(
      () => queue.get(`/requests/${id}?_token=${token}`),
      ({ data }) => data.status === 'SUCCESS',
    )

    expect(req.data).toMatchObject({
      source: 'REQUEST',
      request: { method: 'POST' },
    })

    expect(omit(req.data.request, ['start', 'headers.content-length'])).toMatchSnapshot()
  })

  test('single bulk request with URL without scheme', async () => {
    const res = await queue.post(`/bulk?_token=${token}`, [{
      url: 'postman-echo.com/post',
    }])
    expect(res.data).toHaveLength(1)
    expect(res.data[0]).toHaveProperty('id')
    const id = res.data[0].id

    const req = await waitFor(
      () => queue.get(`/requests/${id}?_token=${token}`),
      ({ data }) => data.status === 'SUCCESS',
    )

    expect(req.data).toMatchObject({
      source: 'REQUEST',
      request: { method: 'POST' },
    })

    expect(omit(req.data.request, ['start', 'headers.content-length'])).toMatchSnapshot()
  })

  test('bulk request with GET method', async () => {
    const res = await queue.post(`/bulk?_token=${token}`, [{
      url: 'https://postman-echo.com/get',
      method: 'GET',
    }])
    expect(res.data).toHaveLength(1)
    expect(res.data[0]).toHaveProperty('id')
    const id = res.data[0].id

    const req = await waitFor(
      () => queue.get(`/requests/${id}?_token=${token}`),
      ({ data }) => data.status === 'SUCCESS',
    )

    expect(req.data).toMatchObject({
      source: 'REQUEST',
      request: { method: 'GET' },
    })

    expect(omit(req.data.request, ['start', 'headers.content-length'])).toMatchSnapshot()
  })

  test('bulk request with query param on URL', async () => {
    const res = await queue.post(`/bulk?_token=${token}`, [{
      url: 'https://postman-echo.com/post?x=1',
    }])
    expect(res.data).toHaveLength(1)
    expect(res.data[0]).toHaveProperty('id')
    const id = res.data[0].id

    const req = await waitFor(
      () => queue.get(`/requests/${id}?_token=${token}`),
      ({ data }) => data.status === 'SUCCESS',
    )

    expect(req.data).toMatchObject({
      source: 'REQUEST',
      request: {
        method: 'POST',
        params: { x: '1' },
        url: 'https://postman-echo.com/post?x=1',
      },
    })
    expect(omit(req.data.request, ['start'])).toMatchSnapshot()

    const resp = await queue.get(`/requests/${id}/response.body?_token=${token}`)
    expect(resp.data).toMatchObject({
      args: { x: '1' },
      url: 'https://postman-echo.com/post?x=1',
    })
  })

  test('bulk request with query param in body', async () => {
    const res = await queue.post(`/bulk?_token=${token}`, [{
      url: 'https://postman-echo.com/post',
      params: { x: '1' },
    }])
    expect(res.data).toHaveLength(1)
    expect(res.data[0]).toHaveProperty('id')
    const id = res.data[0].id

    const req = await waitFor(
      () => queue.get(`/requests/${id}?_token=${token}`),
      ({ data }) => data.status === 'SUCCESS',
    )

    expect(req.data).toMatchObject({
      source: 'REQUEST',
      request: {
        method: 'POST',
        params: { x: '1' },
        url: 'https://postman-echo.com/post?x=1',
      },
    })
    expect(omit(req.data.request, ['start', 'headers.content-length'])).toMatchSnapshot()

    const resp = await queue.get(`/requests/${id}/response.body?_token=${token}`)
    expect(resp.data).toMatchObject({
      args: { x: '1' },
      url: 'https://postman-echo.com/post?x=1',
    })
  })

  test('bulk request with header param', async () => {
    const res = await queue.post(`/bulk?_token=${token}`, [{
      url: 'https://postman-echo.com/post',
      headers: { 'x-custom-header': '123' },
    }])
    expect(res.data).toHaveLength(1)
    expect(res.data[0]).toHaveProperty('id')
    const id = res.data[0].id

    const req = await waitFor(
      () => queue.get(`/requests/${id}?_token=${token}`),
      ({ data }) => data.status === 'SUCCESS',
    )

    expect(req.data).toMatchObject({
      source: 'REQUEST',
      request: {
        method: 'POST',
        headers: { 'x-custom-header': '123' },
        url: 'https://postman-echo.com/post',
      },
    })
    expect(omit(req.data.request, ['start'])).toMatchSnapshot()

    const resp = await queue.get(`/requests/${id}/response.body?_token=${token}`)
    expect(resp.data).toMatchObject({
      headers: { 'x-custom-header': '123' },
      url: 'https://postman-echo.com/post',
    })
  })

  test('bulk request with json body', async () => {
    const res = await queue.post(`/bulk?_token=${token}`, [{
      url: 'https://postman-echo.com/post',
      body: { abc: 1 },
    }])
    expect(res.data).toHaveLength(1)
    expect(res.data[0]).toHaveProperty('id')
    const id = res.data[0].id

    const req = await waitFor(
      () => queue.get(`/requests/${id}?_token=${token}`),
      ({ data }) => data.status === 'SUCCESS',
    )

    expect(req.data).toMatchObject({
      source: 'REQUEST',
      request: {
        method: 'POST',
        headers: { 'content-type': 'application/json' },
        url: 'https://postman-echo.com/post',
      },
    })
    expect(omit(req.data.request, ['start'])).toMatchSnapshot()

    const resp = await queue.get(`/requests/${id}/response.body?_token=${token}`)
    expect(resp.data).toMatchObject({
      headers: { 'content-type': 'application/json' },
      url: 'https://postman-echo.com/post',
      data: { abc: 1 },
      json: { abc: 1 },
    })
  })

  test('bulk request with text body', async () => {
    const res = await queue.post(`/bulk?_token=${token}`, [{
      url: 'https://postman-echo.com/post',
      body: 'helloworld',
    }])
    expect(res.data).toHaveLength(1)
    expect(res.data[0]).toHaveProperty('id')
    const id = res.data[0].id

    const req = await waitFor(
      () => queue.get(`/requests/${id}?_token=${token}`),
      ({ data }) => data.status === 'SUCCESS',
    )

    expect(req.data).toMatchObject({
      source: 'REQUEST',
      request: {
        method: 'POST',
        headers: { 'content-type': 'text/plain' },
        url: 'https://postman-echo.com/post',
      },
    })
    expect(omit(req.data.request, ['start'])).toMatchSnapshot()

    const resp = await queue.get(`/requests/${id}/response.body?_token=${token}`)
    expect(resp.data).toMatchObject({
      headers: { 'content-type': 'text/plain' },
      url: 'https://postman-echo.com/post',
      data: 'helloworld',
    })
  })

  test('bulk request with header, params and body', async () => {
    const res = await queue.post(`/bulk?_token=${token}`, [{
      url: 'https://postman-echo.com/post?y=2',
      params: { x: 1 },
      body: { abc: 1 },
      headers: { 'x-custom-header': '123' },
    }])
    expect(res.data).toHaveLength(1)
    expect(res.data[0]).toHaveProperty('id')
    const id = res.data[0].id

    const req = await waitFor(
      () => queue.get(`/requests/${id}?_token=${token}`),
      ({ data }) => data.status === 'SUCCESS',
    )

    expect(req.data).toMatchObject({
      source: 'REQUEST',
      request: {
        method: 'POST',
        headers: { 'x-custom-header': '123', 'content-type': 'application/json' },
        params: { x: 1, y: '2' },
        hasbody: true,
      },
    })
    expect(qs.parseUrl(req.data.request.url)).toMatchObject({
      url: 'https://postman-echo.com/post',
      query: { x: '1', y: '2' },
    })
    delete req.data.request.url
    expect(omit(req.data.request, ['start'])).toMatchSnapshot()

    const reqBody = await queue.get(`/requests/${id}/request.body?_token=${token}`)
    expect(reqBody.data).toMatchObject({ abc: 1 })

    const respBody = await queue.get(`/requests/${id}/response.body?_token=${token}`)
    expect(respBody.data).toMatchObject({
      headers: { 'x-custom-header': '123' },
      json: { abc: 1 },
      args: { x: '1', y: '2' },
    })
    expect(qs.parseUrl(respBody.data.url)).toMatchObject({
      url: 'https://postman-echo.com/post',
      query: { x: '1', y: '2' },
    })
  })

  test('bulk request with 2 urls', async () => {
    const res = await queue.post(`/bulk?_token=${token}`, [{
      url: 'https://postman-echo.com/post',
    }, {
      url: 'https://postman-echo.com/post',
    }])
    expect(res.data).toHaveLength(2)
    expect(res.data[0]).toHaveProperty('id')
    const id1 = res.data[0].id
    const id2 = res.data[1].id

    const req1 = await waitFor(
      () => queue.get(`/requests/${id1}?_token=${token}`),
      ({ data }) => data.status === 'SUCCESS',
    )

    const req2 = await waitFor(
      () => queue.get(`/requests/${id2}?_token=${token}`),
      ({ data }) => data.status === 'SUCCESS',
    )

    expect(req1.data).toMatchObject({
      source: 'REQUEST',
      request: { method: 'POST' },
    })
    expect(omit(req1.data.request, ['start', 'headers.content-length'])).toMatchSnapshot()

    expect(req2.data).toMatchObject({
      source: 'REQUEST',
      request: { method: 'POST' },
    })
    expect(omit(req2.data.request, ['start', 'headers.content-length'])).toMatchSnapshot()
  })
})
