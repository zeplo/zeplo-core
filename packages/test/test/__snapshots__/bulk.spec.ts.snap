// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bulk.spec bulk request with 2 urls 1`] = `
Object {
  "hasbody": false,
  "headers": Object {},
  "host": "postman-echo.com",
  "method": "POST",
  "params": Object {},
  "path": "/post",
  "scheme": "https",
  "url": "https://postman-echo.com/post",
}
`;

exports[`bulk.spec bulk request with 2 urls 2`] = `
Object {
  "hasbody": false,
  "headers": Object {},
  "host": "postman-echo.com",
  "method": "POST",
  "params": Object {},
  "path": "/post",
  "scheme": "https",
  "url": "https://postman-echo.com/post",
}
`;

exports[`bulk.spec bulk request with GET method 1`] = `
Object {
  "hasbody": false,
  "headers": Object {},
  "host": "postman-echo.com",
  "method": "GET",
  "params": Object {},
  "path": "/get",
  "scheme": "https",
  "url": "https://postman-echo.com/get",
}
`;

exports[`bulk.spec bulk request with header param 1`] = `
Object {
  "hasbody": false,
  "headers": Object {
    "x-custom-header": "123",
  },
  "host": "postman-echo.com",
  "method": "POST",
  "params": Object {},
  "path": "/post",
  "scheme": "https",
  "url": "https://postman-echo.com/post",
}
`;

exports[`bulk.spec bulk request with header, params and body 1`] = `
Object {
  "hasbody": true,
  "headers": Object {
    "content-type": "application/json",
    "x-custom-header": "123",
  },
  "host": "postman-echo.com",
  "method": "POST",
  "params": Object {
    "x": 1,
    "y": "2",
  },
  "path": "/post",
  "scheme": "https",
}
`;

exports[`bulk.spec bulk request with json body 1`] = `
Object {
  "hasbody": true,
  "headers": Object {
    "content-type": "application/json",
  },
  "host": "postman-echo.com",
  "method": "POST",
  "params": Object {},
  "path": "/post",
  "scheme": "https",
  "url": "https://postman-echo.com/post",
}
`;

exports[`bulk.spec bulk request with query param in body 1`] = `
Object {
  "hasbody": false,
  "headers": Object {},
  "host": "postman-echo.com",
  "method": "POST",
  "params": Object {
    "x": "1",
  },
  "path": "/post",
  "scheme": "https",
  "url": "https://postman-echo.com/post?x=1",
}
`;

exports[`bulk.spec bulk request with query param on URL 1`] = `
Object {
  "hasbody": false,
  "headers": Object {},
  "host": "postman-echo.com",
  "method": "POST",
  "params": Object {
    "x": "1",
  },
  "path": "/post",
  "scheme": "https",
  "url": "https://postman-echo.com/post?x=1",
}
`;

exports[`bulk.spec bulk request with text body 1`] = `
Object {
  "hasbody": true,
  "headers": Object {
    "content-type": "text/plain",
  },
  "host": "postman-echo.com",
  "method": "POST",
  "params": Object {},
  "path": "/post",
  "scheme": "https",
  "url": "https://postman-echo.com/post",
}
`;

exports[`bulk.spec single bulk request with URL only 1`] = `
Object {
  "hasbody": false,
  "headers": Object {},
  "host": "postman-echo.com",
  "method": "POST",
  "params": Object {},
  "path": "/post",
  "scheme": "https",
  "url": "https://postman-echo.com/post",
}
`;

exports[`bulk.spec single bulk request with URL without scheme 1`] = `
Object {
  "hasbody": false,
  "headers": Object {},
  "host": "postman-echo.com",
  "method": "POST",
  "params": Object {},
  "path": "/post",
  "scheme": "https",
  "url": "https://postman-echo.com/post",
}
`;
